### PR TITLE
Fix unnecessary update of ra-input-rich-text on edit

### PR DESCRIPTION
**Pull request by roychoo**
Hi,

in ra-input-rich-text,

whenever the text changes, below is the sequence of events happening.
Original text `<p>test</p>`
1. User change text to `<p>test1</p>`
2. `onTextChange` fired
3. `componentDidUpdate` fired
4. update happens as previous prop value is still `<p>test</p>`, but unecessary
5. `onTextChange` fired again.
6. stops there as the prevProp input value is the same as current prop input value

2nd issue is that, it seems quilljs is always adding additional `<p>` after this method
```
this.quill.setContents(
     this.quill.clipboard.convert(this.props.input.value)
)
```
***ONLY*** for the text that has special characters, i.e ```ARTÍCULO 1.- El presente reglamento es de observancia obligatoria``` after you bold ```ARTÍCULO 1```

fixes https://github.com/marmelab/react-admin/issues/3073

### CODE REVIEW

**Comment by djhi** [comment]
Thanks! It seems the tests are failing though

**Comment by roychoo** [comment]
Hi just want to check, did anyone take the opportunity to review the code?

File: test-setup.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 require('raf/polyfill');
+require('mutationobserver-shim');
```
Comment (fzaninotto) [inline test-setup.js]: MutationObserver is used nowhere in the tests, why did you add this?

File: packages/ra-input-rich-text/src/index.spec.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    beforeEach(() => {
+      container = document.createElement('div');
+      document.body.appendChild(container);
+      document.getSelection = () => {
+          return {
+            removeAllRanges: () => {},
+            getRangeAt: function() {}, 
+          };
+      };
+      jest.mock('lodash.debounce', (fn) => fn);
```
Comment (fzaninotto) [inline packages/ra-input-rich-text/src/index.spec.js]: you mocked debounce twice

File: packages/ra-input-rich-text/src/index.spec.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import { RichTextInput } from './index';
+
+let container;
+
+jest.mock('lodash/debounce');
+describe('RichTextInput', () => {
+    beforeEach(() => {
+      container = document.createElement('div');
+      document.body.appendChild(container);
+      document.getSelection = () => {
```
Comment (fzaninotto) [inline packages/ra-input-rich-text/src/index.spec.js]: what's that for? Please comment the code to explain why you need to do that.

File: packages/ra-input-rich-text/src/index.spec.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      jest.mock('lodash.debounce', (fn) => fn);
+    });
+
+    afterEach(() => {
+        document.body.removeChild(container);
+        container = null;
+    });
+
+    it('should call text-change event only once when editing', async () => {
+        const mockFn = jest.fn();
```
Comment (fzaninotto) [inline packages/ra-input-rich-text/src/index.spec.js]: call it handleChange

File: packages/ra-input-rich-text/src/index.spec.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        debounce.mockImplementation(fn => fn);
```
Comment (fzaninotto) [inline packages/ra-input-rich-text/src/index.spec.js]: it seems to be the same implementation as in the beforeEach

File: packages/ra-input-rich-text/src/index.spec.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        <RichTextInput
+            input={{
+              value: '<p>test</p>',
+                onChange: mockFn
+            }}
+            meta={{error: null}} />);
+        const quillNode = await waitForElement(() => {
+            return getByTestId('quill')
+        });
+        quillNode.__quill.setText('test1');
```
Comment (fzaninotto) [inline packages/ra-input-rich-text/src/index.spec.js]: If you're using react-testing-library, this is not the way to go. You should mimick the actions of a real user, not use internal APIs.

File: test-setup.js (hunk 1 above)
Comment (roychoo) [reply test-setup.js]: mutationobserver is used in quilljs

Comment (roychoo) [reply test-setup.js]: i will put in some comments on the ones that require due to quilljs

File: packages/ra-input-rich-text/src/index.spec.js (hunk 6 above)
Comment (roychoo) [reply packages/ra-input-rich-text/src/index.spec.js]: actually, i did tried the normal user way, but quilljs is using `contenteditable` and not input, thus not sure how to simulate a change event. Do let me know if you have any ideas.

File: packages/ra-input-rich-text/src/index.spec.js (hunk 2 above)
Comment (roychoo) [reply packages/ra-input-rich-text/src/index.spec.js]: will remove this

File: packages/ra-input-rich-text/src/index.spec.js (hunk 4 above)
Comment (roychoo) [reply packages/ra-input-rich-text/src/index.spec.js]: done

File: packages/ra-input-rich-text/src/index.spec.js (hunk 5 above)
Comment (roychoo) [reply packages/ra-input-rich-text/src/index.spec.js]: the one in beforeEach is not working, forget to remove it. will remove it

File: test-setup.js (hunk 1 above)
Comment (fzaninotto) [reply test-setup.js]: Then the polyfill should be added by quilljs themselves, or am I missing something?

File: packages/ra-input-rich-text/src/index.spec.js (hunk 6 above)
Comment (fzaninotto) [reply packages/ra-input-rich-text/src/index.spec.js]: You can focus() on the element, then send key events. See https://stackoverflow.com/questions/6139107/programmatically-select-text-in-a-contenteditable-html-element

File: test-setup.js (hunk 1 above)
Comment (roychoo) [reply test-setup.js]: i think in default jest uses jsdom and jsdom do not have support for mutationobserver.

https://github.com/jsdom/jsdom/issues/639

Comment (fzaninotto) [reply test-setup.js]: Right, so you should add that comment to the code so that other developers know why it's included

File: packages/ra-input-rich-text/src/index.spec.js (hunk 6 above)
Comment (roychoo) [reply packages/ra-input-rich-text/src/index.spec.js]: hmm, just realised fireEvent actually works, it was due to the event loop. Had made the changes

File: test-setup.js (hunk 1 above)
Comment (roychoo) [reply test-setup.js]: done.

File: packages/ra-input-rich-text/src/index.spec.js (hunk 4 above)
Comment (djhi) [reply packages/ra-input-rich-text/src/index.spec.js]: Are you sure ? I still see `mockFn`

Comment (roychoo) [reply packages/ra-input-rich-text/src/index.spec.js]: Oh sorry I misunderstood, I thought you meant the test description. Will made the required change in another 3 hours.

Comment (roychoo) [reply packages/ra-input-rich-text/src/index.spec.js]: @djhi done.

**Comment by djhi** [comment]
Thanks!

**Comment by roychoo** [comment]
Thanks for the review and merged
